src/dialogs/sendto/Services/Manager.vala

```diff
@@ -0,0 +1,206 @@
+				BusType.SYSTEM,
+				DBusObjectManagerClientFlags.NONE,
+				"org.bluez",
+				"/",
+				object_manager_proxy_get_type,
+				null
+			);
+
+			object_manager.get_objects().foreach((object) => {
+				object.get_interfaces().foreach((iface) => on_interface_added(object, iface));
+			});
+
+			object_manager.interface_added.connect(on_interface_added);
+
+			object_manager.interface_removed.connect(on_interface_removed);
+
+			object_manager.object_added.connect((object) => {
+				object.get_interfaces().foreach((iface) => on_interface_added(object, iface));
+			});
+
+			object_manager.object_removed.connect((object) => {
+				object.get_interfaces().foreach((iface) => on_interface_removed(object, iface));
+			});
+		} catch (Error e) {
+			critical("Error getting Bluez object manager: %s", e.message);
+		}
+	}
+
+	// TODO: Do not rely on this when it is possible to do it natively in Vala
+	[CCode (cname="bluetooth_device_proxy_get_type")]
+	extern static GLib.Type get_device_proxy_type();
+
+	[CCode (cname="bluetooth_adapter_proxy_get_type")]
+	extern static GLib.Type get_adapter_proxy_type();
+
+	private GLib.Type object_manager_proxy_get_type(DBusObjectManagerClient manager, string object_path, string? interface_name) {
+		if (interface_name == null) return typeof (DBusObjectProxy);
+
+		switch (interface_name) {
+			case "org.bluez.Device1":
+				return get_device_proxy_type();
+			case "org.bluez.Adapter1":
+				return get_adapter_proxy_type();
+			default:
+				return typeof(DBusProxy);
+		}
+	}
+
+	private void register_obex_agentmanager() {
+		try {
+			var connection = GLib.Bus.get_sync(BusType.SESSION);
+			connection.call.begin(
+				"org.bluez.obex",
+				"/org/bluez/obex",
+				"org.bluez.obex.AgentManager1",
+				"RegisterAgent", // TODO: Do we need to worry about unregistering?
+				new Variant("(o)", "/org/bluez/obex/budgie"),
+				null,
+				DBusCallFlags.NONE,
+				-1);
+		} catch (Error e) {
+			critical("Error registering Obex agent manager: %s", e.message);
+		}
+	}
+
+	private void on_interface_added(DBusObject object, DBusInterface iface) {
+		if (iface is Bluetooth.Device) {
+			unowned var device = (Bluetooth.Device) iface;
+			device_added(device);
+		} else if (iface is Bluetooth.Adapter) {
+			unowned var adapter = (Bluetooth.Adapter) iface;
+			has_object = true;
+			((DBusProxy) adapter).g_properties_changed.connect((changed, invalid) => {
+				var discovering = changed.lookup_value("Discovering", VariantType.BOOLEAN);
+				if (discovering != null) {
+					status_discovering();
+				}
+			});
+		}
+	}
+
+	private void on_interface_removed(DBusObject object, DBusInterface iface) {
+		if (iface is Bluetooth.Device) {
+			device_removed((Bluetooth.Device) iface);
+		} else if (iface is Bluetooth.Adapter) {
+			has_object = !get_adapters().is_empty;
+		}
+	}
+
+	public Gee.LinkedList<Bluetooth.Adapter> get_adapters() requires (object_manager != null) {
```

Comment:
Bit too late in the Vala land for me to actually make use of this, but TIL about requires.

Reply:
Yeah, it's pretty nice. If only we'd known about it years ago! xD